python/pyspark/sql/types.py

```diff
@@ -1619,11 +1619,39 @@ def to_arrow_type(dt):
         arrow_type = pa.decimal(dt.precision, dt.scale)
     elif type(dt) == StringType:
         arrow_type = pa.string()
+    elif type(dt) == DateType:
+        arrow_type = pa.date32()
+    elif type(dt) == TimestampType:
+        # Timestamps should be in UTC, JVM Arrow timestamps require a timezone to be read
+        arrow_type = pa.timestamp('us', tz='UTC')
     else:
         raise TypeError("Unsupported type in conversion to Arrow: " + str(dt))
     return arrow_type
 
 
+def _check_dataframe_localize_timestamps(df):
+    """ Convert timezone aware timestamps to timezone-naive in local time
+    """
+    from pandas.api.types import is_datetime64tz_dtype
+    for column, series in df.iteritems():
+        # TODO: handle nested timestamps?
+        if is_datetime64tz_dtype(series.dtype):
+            df[column] = series.dt.tz_convert('tzlocal()').dt.tz_localize(None)
+    return df
+
+
+def _check_series_convert_timestamps_internal(s):
+    """ Convert a tz-naive timestamp in local tz to UTC normalized for Spark internal storage
+    """
+    from pandas.api.types import is_datetime64_dtype
+    # TODO: handle nested timestamps?
```

Comment:
What is the behavior for nested timestamps? Give an example?

Reply:
We should definite add a test to assert what an array<timestamp without time zone> returns

Reply:
I don't believe arrays are supported yet on the python side of things, I plan to look at that next.  Right now it will raise a TypeError

Reply:
Arrays are supported in pyarrow (but perhaps not for timestamps? If that's true could you open a JIRA?), or do you mean something else?

Reply:
Sorry @wesm, I meant on the Spark python side. If a pyspark ArrayType is used a TypeError is raised indicating it is an unsupported type.

Reply:
If it is unsupported, could you also add a negative test case if not existed?

Reply:
@gatorsmile there is a negative test for unsupported types. It ends up the same for any unsupported type (the type is checked in if-elif-else block) so ArrayType is not explicitly tested but would end in the same default case and raise an error. Is that ok with you?